cryostatio/cryostat-agent
# feat(attach): implement basic dynamic attach

Based on #224
Related to #233

This implements basic dynamic attach. The Agent can be launched as a standalone Java process like:

```bash
$ java -jar target/cryostat-agent-0.4.0-SNAPSHOT.jar 1234 # 1234 is another JVM's pid
```

This will cause the Agent process to look up PID `1234` using its Attach providers. If `1234` is found as a valid PID then the Agent process will attach to PID `1234` and attempt to load the Agent's own JAR into that running JVM, which will then bootstrap into the normal Agent launch process. If `1234` is not found then nothing is bootstrapped. Instead of `1234` the wildcard `*` can also be specified, which will lead to the Agent JAR attempting to bootstrap itself into every JVM it finds. Finally, if the PID is not specified or is `0`, then the Agent will check if exactly one candidate JVM is available and attempt to bootstrap into that, failing if there are too many JVM candidates or none.

~~In this mode the Agent still tries to load its configuration entirely from the JVM it is bootstrapping into, which does not really make sense for a late-attaching Agent - the host application probably does not have Cryostat Agent-specific configuration properties if it was not launched with the Cryostat Agent to begin with. This will be addressed some time later in a follow-up PR.~~

Additional late-binding configuration options can be specified to the Agent launcher with command line options:

```bash
$ java -jar target/cryostat-agent-0.4.0-SNAPSHOT.jar \
-Dcryostat.agent.baseuri=http://cryostat.local \
--smartTrigger=[ProcessCpuLoad>0.2]~profile \
@/deployment/app/moreAgentArgs \
1234
```

Try `$ java -jar target/cryostat-agent-0.4.0-SNAPSHOT.jar -h` and see [picocli](https://picocli.info) for a more complete explanation of the available options and their behaviour. These late-binding configuration options are picked up by the Agent launcher process and serialized into a single `String` argument that is passed into the injected Agent instances, which they receive on their `agentmain(String arg)` and then deserialize again. System properties specified with `-D` are set onto the host JVM before the injected Agent attempts to read its configuration values, so from the Agent implementation POV this is identical to setting the `-D` or equivalent environment variable on the host JVM process itself. This mechanism does allow the user who controls the agent injection to override other arbitrary system properties of the host JVM at runtime, however if the user has enough access and ability to inject an agent then they can already exploit this without using the Cryostat Agent specifically.

To test:

```bash
$ ./mvnw clean install
$ gh repo clone andrewazores/quarkus-test
$ cd quarkus-test
$ git checkout agent-dynamic-attach
$ ./build.bash # build a quarkus-test image with the new agent included in the image but not launched with -javaagent
$ ./run.bash # run a quarkus-test container, then after a few seconds, spawn a Cryostat Agent process to dynamically attach to the quarkus server
```

----

Things to investigate:
1. Dynamic attach/agent JAR loading "remotely". If the host application is running in a container that does not already contain the agent JAR, how can we get the agent injected and dynamically attached to the application in the container without having to restart the container? Dynamic attach means the agent loading doesn't require a restart, but adding a new mounted volume to the container might. The attach mechanism, at least the one used in this PR, allows for attachment to a remote VM, but requesting the VM to load the agent requires a JAR path local to the VM's filesystem. Is there some way to get the VM to load from a `ByteArrayInputStream` or something so we can send it the JAR contents directly?
2. or 1b - is it feasible to rewrite parts of the agent to not actually rely on being part of the host application directly, but instead use only the local management agent (JMX)? Would also need to investigate how to ensure that's kept locked down and never exposed to any other processes, at least none outside of the container, if the agent launcher is the one turning that feature on. If this can be done then actually the agent would not need to inject itself into VMs but can run alongside them as a separate process and use local management RMI. This seems a bit hairy and is probably best avoided however. Perhaps that's best suited for another component that is not an agent, but rather a sidecar collector process, which may not implement all of the same features as the agent.
3. Watch mode? Currently the agent launcher will execute once, attempting to inject itself into one or more VMs, and then exit. If the user specified the `*` PID wildcard, it might also be useful to have a `--watch` option where the agent launcher becomes a daemon, injecting agent instances into new JVMs as they appear.
